src/Tutor.Web/Controllers/Analytics/AnalyticsController.cs

```diff
@@ -49,8 +49,13 @@ public ActionResult<List<KcStatisticsDto>> GetKcStatistics([FromQuery] int unitI
         var unit = _knowledgeUnitRepository.Get(unitId);
         List<int> learnerIds = null;
         if(groupId != 0) learnerIds = _learnerRepository.GetByGroupId(groupId).Select(l => l.Id).ToList();
+        List<int> kcIds = unit.KnowledgeComponents.Select(kc => kc.Id).ToList();
+
+        var eventQuery = _eventStore.Events.Where(e => kcIds.Contains(e.RootElement.GetProperty("KnowledgeComponentId").GetInt32()));
```

Comment:
Since this controller is concerned with too many responsibilities, I'm thinking what would be a better approach to segregating the code.

The current IEventStore is generic which is good, but maybe we should be thinking about an IKnowledgeComponentEventStore that offers a method that encapsulates lines 54-57, accepting the ids and return a list of events. It's implementation can work with the IEventStore.

What do you think?

Reply:
I would probably just encapsulate these lines in a repository instead, that seems like the best place for all code that uses IEventStore to me (the same way any code that uses EF Core is encapsulated in repositories). But I definitely do agree that the AnalyticsController needs some refactoring.